src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +499,137 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
+out:
+#endif
+    return status;
+}
+
+static void uct_cuda_copy_sync_memops(const void *address)
+{
+#if HAVE_CUDA_FABRIC
+    unsigned flags;
+#else
+    unsigned value = 1;
+#endif
+
+#if HAVE_CUDA_FABRIC
+    /* Synchronize future DMA operations for all memory types */
+    UCT_CUDADRV_FUNC_LOG_WARN(cuCtxGetFlags(&flags));
+    if (!(flags & CU_CTX_SYNC_MEMOPS)) {
+        UCT_CUDADRV_FUNC_LOG_WARN(cuCtxSetFlags(CU_CTX_SYNC_MEMOPS));
+    }
+#else
+    /* Synchronize for DMA for legacy memory types*/
+    UCT_CUDADRV_FUNC_LOG_WARN(cuPointerSetAttribute(&value,
+                CU_POINTER_ATTRIBUTE_SYNC_MEMOPS, (CUdeviceptr)address));
+#endif
+
+}
+
 static ucs_status_t
 uct_cuda_copy_md_query_attributes(uct_cuda_copy_md_t *md, const void *address,
                                   size_t length, ucs_memory_info_t *mem_info)
 {
 #define UCT_CUDA_MEM_QUERY_NUM_ATTRS 4
-    CUmemorytype cuda_mem_mype = (CUmemorytype)0;
+    CUmemorytype cuda_mem_type = CU_MEMORYTYPE_HOST;
     uint32_t is_managed        = 0;
     CUdevice cuda_device       = -1;
     CUcontext cuda_mem_ctx     = NULL;
     CUpointer_attribute attr_type[UCT_CUDA_MEM_QUERY_NUM_ATTRS];
     void *attr_data[UCT_CUDA_MEM_QUERY_NUM_ATTRS];
+    ucs_memory_type_t vmm_mem_type;
     CUdeviceptr base_address;
     size_t alloc_length;
     size_t total_bytes;
     int32_t pref_loc;
+    unsigned is_vmm;
     CUresult cu_err;
+    ucs_status_t status;
 
     attr_type[0] = CU_POINTER_ATTRIBUTE_MEMORY_TYPE;
-    attr_data[0] = &cuda_mem_mype;
+    attr_data[0] = &cuda_mem_type;
     attr_type[1] = CU_POINTER_ATTRIBUTE_IS_MANAGED;
     attr_data[1] = &is_managed;
     attr_type[2] = CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL;
     attr_data[2] = &cuda_device;
     attr_type[3] = CU_POINTER_ATTRIBUTE_CONTEXT;
     attr_data[3] = &cuda_mem_ctx;
 
-    cu_err = cuPointerGetAttributes(ucs_static_array_size(attr_data), attr_type,
-                                    attr_data, (CUdeviceptr)address);
-    if ((cu_err != CUDA_SUCCESS) || (cuda_mem_mype != CU_MEMORYTYPE_DEVICE)) {
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuPointerGetAttributes(
+                ucs_static_array_size(attr_data), attr_type, attr_data,
+                (CUdeviceptr)address));
+    if (status != UCS_OK) {
+        /* pointer not recognized */
+        return status;
+    }
+
+    status = uct_cuda_copy_detect_vmm((void*)address, &is_vmm, &vmm_mem_type);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    if (is_vmm) {
+        if (vmm_mem_type == UCS_MEMORY_TYPE_HOST) {
+            /* Overwrite cuda_mem_type if EGM detected to allow cuda_ipc to
+             * access the memory type. TODO: Remove this logic after cuda_ipc
+             * can support host memory access */
+            cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+        } else if (vmm_mem_type == UCS_MEMORY_TYPE_CUDA) {
+            /* cuda_mem_type should already be CU_MEMORYTYPE_DEVICE */
+        } else {
+            return UCS_ERR_INVALID_ADDR;
```

Comment:
seems UCS_MEMORY_TYPE_UNKNOWN is a considered an invalid result.
so maybe uct_cuda_copy_detect_vmm can return "int" to indicate if the memory was detected as vmm or not, and the value of &vmm_mem_type would indicate success/failure of the detailed VMM memory detection